src/darker/formatters/black_formatter.py

```diff
@@ -0,0 +1,205 @@
+
+In examples below, a simple two-line snippet is used.
+The first line will be reformatted by Black, and the second left intact::
+
+    >>> from pathlib import Path
+    >>> from unittest.mock import Mock
+    >>> src = Path("dummy/file/path.py")
+    >>> src_content = TextDocument.from_lines(
+    ...     [
+    ...         "for i in range(5): print(i)",
+    ...         'print("done")',
+    ...     ]
+    ... )
+
+First, `BlackFormatter.run` uses Black to reformat the contents of a given file.
+Reformatted lines are returned e.g.::
+
+    >>> from darker.formatters.black_formatter import BlackFormatter
+    >>> dst = BlackFormatter().run(src_content, src)
+    >>> dst.lines
+    ('for i in range(5):', '    print(i)', 'print("done")')
+
+See :mod:`darker.diff` and :mod:`darker.chooser`
+for how this result is further processed with:
+
+- :func:`~darker.diff.diff_and_get_opcodes`
+  to get a diff of the reformatting
+- :func:`~darker.diff.opcodes_to_chunks`
+  to split the diff into chunks of original and reformatted content
+- :func:`~darker.chooser.choose_lines`
+  to reconstruct the source code from original and reformatted chunks
+  based on whether reformats touch user-edited lines
+
+"""
+
+from __future__ import annotations
+
+import logging
+from typing import TYPE_CHECKING, TypedDict
+
+from black import FileMode as Mode
+from black import (
+    TargetVersion,
+    format_str,
+    parse_pyproject_toml,
+    re_compile_maybe_verbose,
+)
+
+from darker.files import find_pyproject_toml
+from darker.formatters.base_formatter import BaseFormatter, HasConfig
+from darker.formatters.formatter_config import (
+    BlackCompatibleConfig,
+    read_black_compatible_cli_args,
+    validate_target_versions,
+)
+from darkgraylib.config import ConfigurationError
+from darkgraylib.utils import TextDocument
+
+if TYPE_CHECKING:
+    from argparse import Namespace
+    from pathlib import Path
+    from typing import Pattern
+
+__all__ = ["Mode"]
+
+logger = logging.getLogger(__name__)
+
+
+class BlackModeAttributes(TypedDict, total=False):
+    """Type definition for items accepted by ``black.Mode``."""
+
+    target_versions: set[TargetVersion]
+    line_length: int
+    string_normalization: bool
+    is_pyi: bool
+    magic_trailing_comma: bool
+    preview: bool
+
+
+class BlackFormatter(BaseFormatter, HasConfig[BlackCompatibleConfig]):
+    """Black code formatter plugin interface."""
+
+    config: BlackCompatibleConfig  # type: ignore[assignment]
+
+    config_section = "tool.black"
+
+    def read_config(self, src: tuple[str, ...], args: Namespace) -> None:
+        """Read Black configuration from ``pyproject.toml``.
+
+        :param src: The source code files and directories to be processed by Darker
+        :param args: Command line arguments
+
+        """
+        config_path = args.config or find_pyproject_toml(src)
+        if config_path:
+            self._read_config_file(config_path)
+        self._read_cli_args(args)
+
```

Comment:
Helper method would help cleanup lines 100->125
def _add_to_config_if_exists(self, raw_config, keys):
    for name in keys:
        if name in raw_config:
            self.config[name] = raw_config[name]

Reply:
It's neat, but we then lose static type checking. Running Mypy on such code will give us
```python
error: TypedDict key must be a string literal; expected one of (
    "config", "exclude", "extend_exclude", "force_exclude", "target_version", ...
)  [literal-required]
```